src/adr/0036-against-use-of-conditional-statements-in-guice-modules.md

```diff
@@ -0,0 +1,110 @@
+# 36. Against the use of conditional statements in Guice modules
+
+Date: 2019-12-29
+
+## Status
+
+Accepted (lazy consensus)
+
+## Context
+
+James products rely historically on Spring for dependency injection. It doesn't use last major Spring version (4.x instead of 5.x).
+James uses Spring in a way that enables overriding any class via a configuration file thus endangering overall correctness by giving too much 
+power to the user.
+
+James propose several implementations for each of the interfaces it defines. The number of possible combinations of implementations is thus really high (like factorial(n) with n > 10). It makes it unpractical to run tests for each possible interface
+combination. We rather run integration tests for combinations that we decide brings the more value to the users. By overriding any class, Spring defeats 
+this testing logic. Knowing which component implementation combines well with which other one is not intuitive to users,
+hard to document and test.
+
+We thus decided to rely on Guice. Components are defined by code in a static fashion. Overriding components requires 
+explicit code modification and recompilation, thus warning the user about the impact of the choices he does, and lowering 
+the project's responsibility.
+
+With Guice we expose only supported, well tested combinations of components, thus addressing the combination issue.
+
+Instead of having a single big application being able to instantiate each and every component application, we have 
+several products defining their dependencies in a minimalistic way, relying only on the components implementation that 
+are needed.
+
+Here is the list of products we provide:
+
+ - In-Memory: A memory based James server, mainly for testing purposes
+ - Distributed James: A scalable James server, storing data in various data stores. Cassandra is used for metadata, 
+ ElasticSearch for search, RabbitMQ for messaging, and ObjectStorage for blobs.
+ - Cassandra: An implementation step toward Distributed James. It does not include messaging and ObjectStorage and should not be run in a cluster way but is still relevant for good performance.
```

Comment:
can we state that it has chance to be deprecated/removed soon?

Reply:
is it the same soon as last year?